extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -0,0 +1,243 @@
+   */
+  constructor(ampdoc) {
+    /** @const {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampdoc = ampdoc;
+
+    /** @const {!Array<BindingDef>} */
+    this.bindings_ = [];
+
+    /** @const {!Object} */
+    this.scope_ = Object.create(null);
+
+    /** @const {!../../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = vsyncFor(ampdoc.win);
+
+    /** @const {!Array<string>} */
+    this.protocolWhitelist_ = ['http', 'https'];
+
+    this.ampdoc.whenBodyAvailable().then(body => {
+      this.bindings_ = this.scanForBindings_(body);
+
+      // Trigger verify-only digest in development.
+      if (getMode().development) {
+        this.digest_(true);
+      }
+    });
+  }
+
+  /**
+   * @param {!Object} state
+   * @param {boolean=} opt_skipDigest
+   */
+  setState(state, opt_skipDigest) {
+    Object.assign(this.scope_, state);
+
+    if (!opt_skipDigest) {
+      this.digest_();
+    }
+  }
+
+  /**
+   * Scans children for attributes that conform to bind syntax and returns
+   * all bindings.
+   * @param {!Element} body
+   * @return {!Array<BindingDef>}
+   * @private
+   */
+  scanForBindings_(body) {
+    const bindings = [];
+    const elements = body.getElementsByTagName('*');
+    for (let i = 0; i < elements.length; i++) {
+      const el = elements[i];
+      const attributes = el.attributes;
+      for (let j = 0; j < attributes.length; j++) {
+        const binding = this.bindingForAttribute_(attributes[j], el);
+        if (binding) {
+          bindings.push(binding);
+        }
+      }
+    }
+    return bindings;
+  }
+
+  /**
+   * Returns a struct representing the binding corresponding to the
+   * attribute param, if applicable.
+   * @param {!Attr} attribute
+   * @param {!Element} element
+   * @return {?BindingDef}
+   * @private
+   */
+  bindingForAttribute_(attribute, element) {
+    // TODO: Disallow binding to blacklisted attributes.
+
+    if (attribute.name.length <= 2) {
+      return null;
+    }
+    const name = attribute.name;
+    if (name.charAt(0) === '[' && name.charAt(name.length - 1) === ']') {
+      return {
+        property: name.substr(1, name.length - 2),
+        expression: attribute.value,
+        element,
+      };
+    }
+  }
+
+  /**
+   * Schedules a vsync task to reevaluate all binding expressions.
+   * @param {boolean=} opt_verifyOnly
+   * @private
+   */
+  digest_(opt_verifyOnly) {
+    for (let i = 0; i < this.bindings_.length; i++) {
+      const binding = this.bindings_[i];
+
+      /** {!VsyncTaskSpecDef} */
+      const task = {
+        measure: state => {
+          state.result = evaluateBindExpr(binding.expression, this.scope_);
+        },
+        mutate: state => {
+          if (opt_verifyOnly) {
+            this.verifyBinding_(binding, state.result);
+          } else {
+            this.applyBinding_(binding, state.result);
+          }
+        },
+      };
+      this.vsync_.run(task, {});
```

Comment:
Instead of registering N vsync tasks, could we register one and run the loop there?